Add ws66i core integration

ssaenger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I've added a new core integration called ws66i. This is based off of the https://www.home-assistant.io/integrations/monoprice integration but uses an updated amplifier that enables control via a different protocol (Telnet). Created and added new dependency https://pypi.org/project/pyws66i/ to manage the device interface.

WS66i product page: https://www.soundavo.com/products/ws-66i

This is my first time working with Python. I'm a Bluetooth Software Engineer and I program in C mainly.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19294

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description. NOT SURE WHAT IT IS ASKING
- [ ] Untested files have been added to `.coveragerc`. ALL MY FILES HAVE BEEN TESTED

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ssaenger committed:
Add ws66i core integration

ssaenger commented:
@etsinko, @OnFreund, can either of you take a look at my PR?

OnFreund commented:
Looks like a cool device (super expensive though).
It's going to take me a while to go over this, and I wouldn't want to be a bottleneck, but I will start with one comment - translations happen automatically, you shouldn't commit the language files.

ssaenger commented:
I wasn't sure about translations; I thought they were created manually. I included them because it is the same as the monoprice integration, except I swapped "port" with "ip address". Those will have to be checked... Or I just remove them altogether.

OnFreund commented:
You should remove all the translation files (but keep `strings.json`

ssaenger committed:
Remove all ws66i translations

ssaenger committed:
Update ws66i unit tests to meet minimum code coverage

ssaenger commented:
Can we run automated tests to verify my PR meets all requirements?

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,70 @@
+        except CannotConnect as err:
+            raise ConfigEntryNotReady from err
+
+    undo_listener = entry.add_update_listener(_update_listener)
```
```suggestion
    entry.async_on_unload(entry.add_update_listener(_update_listener))
```

ssaenger replied:
Done. Removed all references to UNDO_UPDATE_LISTENER.

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,70 @@
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
+        WS66I_OBJECT: ws66i,
+        UNDO_UPDATE_LISTENER: undo_listener,
```
```suggestion
```

ssaenger replied:
Done. Removed all references to UNDO_UPDATE_LISTENER.

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,70 @@
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN][entry.entry_id][WS66I_OBJECT].close()
+        hass.data[DOMAIN][entry.entry_id][UNDO_UPDATE_LISTENER]()
```
```suggestion
```

ssaenger replied:
Done. Removed all references to UNDO_UPDATE_LISTENER.

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,149 @@
+    ws66i = get_ws66i(data[CONF_IP_ADDRESS])
+    try:
+        await hass.async_add_executor_job(ws66i.open)
+    except ConnectionError as err:
+        raise CannotConnect from err
```
No need to create `CannotConnect` since the library already has an exception for this. You can use `ConnectionError` directly instead

ssaenger replied:
I removed the try block inside this method. It looks like my calling method catches the exception.

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,149 @@
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Monoprice 6-Zone Amplifier."""
```
stale docstring

ssaenger replied:
Done

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,149 @@
+    @core.callback
+    def async_get_options_flow(config_entry):
+        """Define the config flow to handle options."""
+        return MonopriceOptionsFlowHandler(config_entry)
```
stale name

ssaenger replied:
Done

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,149 @@
+    return key
+
+
+class MonopriceOptionsFlowHandler(config_entries.OptionsFlow):
```
stale name

ssaenger replied:
Done

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,149 @@
+
+
+class MonopriceOptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle a Monoprice options flow."""
```
stale docstring

ssaenger replied:
Done

bdraco commented on tests/components/ws66i/__init__.py:
```diff
@@ -0,0 +1 @@
+"""Tests for the monoprice component."""
```
stale docstring

ssaenger replied:
Done

bdraco commented on tests/components/ws66i/test_config_flow.py:
```diff
@@ -0,0 +1,119 @@
+"""Test the Monoprice 6-Zone Amplifier config flow."""
```
stale docstring

ssaenger replied:
Done

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,70 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Soundavo WS66i 6-Zone Amplifier from a config entry."""
+
+    # double negative to handle absence of value
+    first_run = not bool(entry.data.get(CONF_NOT_FIRST_RUN))
+    if first_run:
+        hass.config_entries.async_update_entry(
+            entry, data={**entry.data, CONF_NOT_FIRST_RUN: True}
+        )
+    else:
+        # Need to verify we can connect
+        try:
+            _ = await validate_input(hass, entry.data)
+        except CannotConnect as err:
+            raise ConfigEntryNotReady from err
```
Check this in the config flow before creating the entry.  Don't create the entry if the input isn't valid

ssaenger replied:
I do validate user input in the config flow. I don't recall why I do it here in the async_setup_entry() method. I'm not sure if I added in response to some testing of mine...

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the WS66i 6-zone amplifier platform."""
+    addr = config_entry.data[CONF_IP_ADDRESS]
+
+    # Open up a connection with the WS66i
+    ws66i = hass.data[DOMAIN][config_entry.entry_id][WS66I_OBJECT]
+    try:
+        await hass.async_add_executor_job(ws66i.open)
+    except ConnectionError as err:
+        raise ConfigEntryNotReady from err
+
```
Do this in `__init__`'s `async_setup_entry` so you don't get to platform setup if there is a ConnectionError

ssaenger replied:
ws66i.open() creates a connection with the amplifier. I call this here and keep it open until the component is unloaded. However, I also call ws66i.open in my __init__'s async_setup_entry as well to verify that a connection can be made, and then I close it.

bdraco replied:
This needs to be done in `__init__` as `ConfigEntryNotReady` does not work in a platform

ssaenger committed:
Update ws66i based on @bdraco review

ssaenger commented:
Thanks for the review @bdraco. I pushed changes based on it.

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,62 @@
+    """Set up Soundavo WS66i 6-Zone Amplifier from a config entry."""
+
+    # double negative to handle absence of value
+    first_run = not bool(entry.data.get(CONF_NOT_FIRST_RUN))
+    if first_run:
+        hass.config_entries.async_update_entry(
+            entry, data={**entry.data, CONF_NOT_FIRST_RUN: True}
+        )
```
Do the first run when the device is setup in the config entry so we don't have to do it here and then we don't have to keep track of this

ssaenger replied:
I got rid of it. It was used to know whether to update entities before registration in order to know which entities to enable/disable. I made some improvements there where the "first run" is not needed.

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,62 @@
+            entry, data={**entry.data, CONF_NOT_FIRST_RUN: True}
+        )
+    else:
+        # Need to verify we can connect
+        try:
+            _ = await validate_input(hass, entry.data)
+        except ConnectionError as err:
+            raise ConfigEntryNotReady from err
```
It looks like you are creating the object twice.  Once here and once below.

We should only create it once.

`validate_input` is purpose built for the config flow.  If you want to reuse it, break it apart into multiple functions so it only does what you need here.

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+    first_run = hass.data[DOMAIN][config_entry.entry_id][FIRST_RUN]
+    async_add_entities(entities, first_run)
+
+    def close(event):
+        """Close the Telnet connection."""
+        ws66i.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close)
```
Move this to `__init__` as well

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return self._zone_id < 20 or self._update_success
+
+    @property
+    def device_info(self):
+        """Return device info for this device."""
+        return {
+            "identifiers": {(DOMAIN, self.unique_id)},
+            "name": self.name,
+            "manufacturer": "Soundavo",
+            "model": "WS66i 6-Zone Amplifier",
+        }
```
This can be set as `self._attr_device_info = DeviceInfo({....}`

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+            "model": "WS66i 6-Zone Amplifier",
+        }
+
+    @property
+    def unique_id(self):
+        """Return unique ID for this device."""
+        return self._unique_id
```
This can be set as `self._attr_unique_id`

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+        """Return unique ID for this device."""
+        return self._unique_id
+
+    @property
+    def name(self):
+        """Return the name of the zone."""
+        return self._name
```
This can be set as `self._attr_name`

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+        DOMAIN,
+        SERVICE_SNAPSHOT,
+        async_service_handle,
+        schema=cv.make_entity_service_schema({}),
```
```suggestion
        schema={},
```

bdraco replied:
Use entity services and pass a `dict` instead

https://developers.home-assistant.io/docs/dev_101_services?_highlight=entity&_highlight=services#entity-services

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+        DOMAIN,
+        SERVICE_RESTORE,
+        async_service_handle,
+        schema=cv.make_entity_service_schema({}),
```
```suggestion
        schema={},
```

bdraco replied:
Use entity services and pass a `dict` instead

https://developers.home-assistant.io/docs/dev_101_services?_highlight=entity&_highlight=services#entity-services

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,278 @@
+
+@core.callback
+def _get_sources(config_entry):
+    if CONF_SOURCES in config_entry.options:
+        data = config_entry.options
+    else:
+        data = config_entry.data
+    return _get_sources_from_dict(data)
```
Store these in `.options` when you create the config entry so you don't have to check both places

bdraco reviewed (changes requested):
Please see comments above

ssaenger commented:
Thanks for the review. I will add suggested changes this week.

ssaenger committed:
General improvements after 2nd PR review

ssaenger commented:
I incorporated @bdraco 's suggestions and added an extra verification to determine that the correct connection was made. I changed the way entities are registered. Previously it would register all possible entities and disable the ones that are not usable. Now I only register usable entities. This makes the end-user experience better because the user's screen isn't populated with 18 different entities when most likely they would only use 6 of them.

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,59 @@
+        await hass.async_add_executor_job(ws66i.open)
+    except ConnectionError:
+        # Amplifier is probably turned off
+        _LOGGER.warning("Could not connect to WS66i Amp. Is it off?")
```
Raise ConfigEntryNotReady(msg) instead

ssaenger replied:
Why should ConfigEntryNotReady be raised here? When should I return False in async_setup_entry()?

The name of the exception to me sounds like an exception that implies the information in the config entry is not ready to be used. Is that thought right?

bdraco replied:
It will cause the integration to retry setup later in the hopes the device has been turned on. Otherwise a restart will be required to get it back up and going

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+    ws66i = get_ws66i(data[CONF_IP_ADDRESS])
+    await hass.async_add_executor_job(ws66i.open)
+    # no exception. run a simple test to make sure we opened correct port
+    ret_val = await hass.async_add_executor_job(ws66i.zone_status, 11)
```
What is 11 here?  Please make this a constant so it shows intent

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,217 @@
+    @property
+    def entity_registry_enabled_default(self):
+        """Return if the entity should be enabled when first added to the entity registry."""
+        return True
```
This is the default. You can drop the property here

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,217 @@
+    @property
+    def supported_features(self):
+        """Return flag of media commands that are supported."""
+        return SUPPORT_WS66I
```
You can set this as _attr_supported_features above __init__

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,217 @@
+    @property
+    def source_list(self):
+        """List of available input sources."""
+        return self._source_names
```
You can use an _attr here

bdraco commented on homeassistant/components/ws66i/strings.json:
```diff
@@ -0,0 +1,40 @@
+        "title": "Connect to the device",
+        "data": {
+          "ip_address": "[%key:common::config_flow::data::ip%]",
+          "source_1": "Name of source #1",
```
We shouldn't require theses to setup the device. If you can't detect them automatically please move them to an options flow only.  We only want the data needed to connect in the base config flow setup

ssaenger committed:
Disable entities if amp shutoff, set default source names, set 30sec polling

ssaenger commented:
- Incorporated @bdraco 's changes.
- Increased poll time from 10 (default) to 30 seconds. I felt like 10 seconds is too fast to poll the state, and if user makes a change to a zone in home assistant, the amp state is updated immediately.
- Set default zone names. If this was not done, user would not be able to select zones until user configured it through an options flow.
- Disable polling and set device state to unavailable if amplifier turns off at anytime. This is to prevent logs from filling with error messages, and to save on resources.

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,214 @@
+        self._state = None
+        self._volume = None
+        self._source = None
+        self._mute = None
```
```suggestion
        self._attr_is_volume_muted = None
```

bdraco replied:
There is a complete list of ones that can be converted here

https://github.com/home-assistant/core/blob/dev/homeassistant/components/media_player/__init__.py#L400

ssaenger replied:
Thanks. This is useful.

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,214 @@
+        # successfully retrieved zone state
+        self._state = STATE_ON if new_state.power else STATE_OFF
+        self._volume = new_state.volume
+        self._mute = new_state.mute
```
```suggestion
        self._attr_is_volume_muted = new_state.mute
```

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,214 @@
+            return None
+        return self._volume / 38.0
+
+    @property
+    def is_volume_muted(self):
+        """Boolean if volume is currently muted."""
+        return self._mute
+
```
```suggestion
```

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,214 @@
+    # Zones 11 - 16 are the master amp
+    # Zones 21,31 - 26,36 are the daisy-chained amps
+    entities = []
+    for i in range(1, 4):
```
Please show the intent of this code by naming this something other than `i`.

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,214 @@
+                break
+
+        _LOGGER.info("Detected amp %d at ip %s", i, ip_addr)
+        for j in range(1, 7):
```
Please show the intent of this code by naming this something other than `j`.

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,56 @@
+        options = {CONF_SOURCES: INIT_OPTIONS_DEFAULT}
+        hass.config_entries.async_update_entry(entry, options=options)
+
+    entry.async_on_unload(entry.add_update_listener(_update_listener))
```
Move this after the `ConfigEntryNotReady` so it doesn't have to get removed and recreated each time setup is retried

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,56 @@
+        """Close the Telnet connection to the amplifier."""
+        ws66i.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close)
```
```suggestion
    entry.async_on_unload(hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close))
```

Otherwise this will leak when the integration is unloaded

bdraco commented on homeassistant/components/ws66i/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "requirements": ["pyws66i==0.1"],
+  "codeowners": ["@ssaenger"],
+  "config_flow": true,
+  "iot_class": "local_polling"
```
Can you discover this via DHCP ?

https://developers.home-assistant.io/docs/creating_integration_manifest?_highlight=dhcp#dhcp

ssaenger replied:
That is something I want to investigate as it seems like it can, but because I am new to HA and python, I am using the monoprice integration as a template for this one (since they are very similar), and that one can't do DHCP discovery. I want to get this approved and added before tackling DHCP.

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,143 @@
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None):
+        """Manage the options."""
+        if user_input is not None:
+            return self.async_create_entry(
+                title="Source Names",
+                data={CONF_SOURCES: _sources_from_config(user_input)},
+            )
+
+        # Fill form with previous source names
+        previous_sources = self.config_entry.options[CONF_SOURCES]
+        options = {
+            _key_for_source(idx + 1, source, previous_sources): str
+            for idx, source in enumerate(SOURCES)
+        }
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema(options),
+        )
```
Its a bit unusual to not be able to get the names of the sources from the device

ssaenger replied:
The amplifier doesn't provide names of sources. So I set them to "Source 1", "Source 2", etc. And I prefill the options flow with those names, unless user changes them, then I prefill with those. Also, if I don't set default names, sources won't be able to be selected when controlling the amp via HA until they are set.

ssaenger committed:
Add _attr_ and change async_on_unload

ssaenger commented:
Incorporated @bdraco suggestions.

bdraco commented on homeassistant/components/ws66i/config_flow.py:
```diff
@@ -0,0 +1,143 @@
+    # Test on FIRST_ZONE because this zone will always be valid
+    ret_val = await hass.async_add_executor_job(ws66i.zone_status, FIRST_ZONE)
+    if ret_val is None:
+        ws66i.close()
+        raise ConnectionError("Not a valid WS66i connection")
+
+    # Validation done. No issues. Close the connection
+    ws66i.close()
```
Is `.close` safe to run in the event loop?

ssaenger replied:
Is a method considered safe to run on the event loop if it is not blocking? .close calls Telnet.close. The documentation there just says "close the connection". Looking for more information online, close will finish sending data on the pipeline and send EOF. It doesn't not wait for a response. To me, this sounds like it is not blocking, so safe. However the HA documentation says no I.O allowed. So is the following acceptable?
`    async def async_close(event):
        """Close the Telnet connection to the amplifier."""
        await hass.async_add_executor_job(ws66i.close)
        entry.async_on_unload(hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_close))
`
In my testing it calls close when HA shutsdown.

bdraco replied:
Traced the code. Looks like it should be safe

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,199 @@
+    # List of custom names
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
+
+    return [source_id_name, source_name_id, source_names]
```
This would be much cleaner if you used a dataclass 

https://docs.python.org/3/library/dataclasses.html

ssaenger replied:
How would that look like? Would I be returning a class object from this function? I would have to initialize a dataclass and then access those fields using "dataclasses.fields()"

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,199 @@
+    # Zones 11 - 16 are the master amp
+    # Zones 21,31 - 26,36 are the daisy-chained amps
+    entities = []
+    for amp_num in range(1, 4):
+
+        if amp_num > 1:
+            # Don't add entities that aren't present
+            status = await hass.async_add_executor_job(
+                ws66i.zone_status, (amp_num * 10 + 1)
+            )
+            if status is None:
+                break
+
+        _LOGGER.info("Detected amp %d at ip %s", amp_num, ip_addr)
+        for zone_num in range(1, 7):
+            zone_id = (amp_num * 10) + zone_num
+            entities.append(Ws66iZone(ws66i, sources, config_entry.entry_id, zone_id))
```
Move this into its own function and wrap it in `hass.async_add_executor_job` so you aren't creating an executor job for each loop

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,199 @@
+class Ws66iZone(MediaPlayerEntity):
+    """Representation of a WS66i amplifier zone."""
+
+    def __init__(self, ws66i, sources, namespace, zone_id):
```
```suggestion
    def __init__(self, ws66i, sources, entry_id, zone_id):
```

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,199 @@
+        self._source_name_id = sources[1]
+        # ordered list of all source names
+        self._attr_source_list = sources[2]
+        self._attr_unique_id = f"{namespace}_{self._zone_id}"
```
```suggestion
        self._attr_unique_id = f"{entry_id}_{self._zone_id}"
```

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,199 @@
+            # End-user should turn on Amp and reload integration
+            _LOGGER.warning("Zone %d not detected. Is it turned off?", self._zone_id)
+            self._attr_available = False
+            self._attr_should_poll = False
```
```suggestion
```

Generally we let it keep polling and recover when it turns back on

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,199 @@
+        new_state = self._ws66i.zone_status(self._zone_id)
+        if not new_state:
+            # End-user should turn on Amp and reload integration
+            _LOGGER.warning("Zone %d not detected. Is it turned off?", self._zone_id)
```
https://developers.home-assistant.io/docs/integration_quality_scale_index/#silver-

Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.

You can avoid all the complexity of handling that by using the coordinator https://developers.home-assistant.io/docs/integration_fetching_data?_highlight=datau#coordinated-single-api-poll-for-data-for-all-entities

ssaenger replied:
I don't understand how coordinators work. The HA documentation doesn't go into detail how it works, so I have to look at other integration's examples to figure it out. I make an observation in an example for instance, and compare it with the others. If I see a pattern among them, I try to understand why they are doing it the way they are. Most seem to create the DataUpdateCoordinator in the __init__.py file and the CoordinatorEntity in the platform file. Not everyone follows this convention, but I assume it's the preferred way to organize things. There is always 1 DataUpdateCorrdinator and 1 or more CoordinatorEntitys. I need at least 6 entities to represent each of the zones on the WS66i, so I assume I want to create 6 of them. But I don't understand how those 6 entities get updated. 

Before coordinators, I would create my 6 entities and call `async_add_entities()` in my platform file. HA uses this probably to iterate through all entities with polling enabled, calls it's `update()` method, and then reads each of the entities attributes. This makes sense; I understand how HA retrieves my entities data... But with coordinators, it's not clear exactly what it does. It looks like I still add my entities (but now which inherits from CoordinatorEntity) through the `async_add_entities() method`. So, do each of my entities `update()` method still get called? If so, then what is the point of the DataUpdateCoordinator, which the documentation states helps me update all my entities from a single endpoint. In fact, I see from the example in the docs that a coordinator is created, then `async_config_entry_first_refresh()` is called before any of the entities are created! So how do the entities get updated from this call?!

I've tried to rationalize that HA will call my coordinator's `_async_update_data()`, but I don't see how HA knows to call it. But if it somehow knows and does call it, then what I would think it would do it look at each entity that is manages, then call each of those entities `update()` methods. Yet this is what I don't see. It would make sense if it worked this way, but that is not the case. In fact by looking at the other integrations examples, I don't see how the coordinator even is aware of my entities, so I am still confused how they get updated. It only looks like my entities are aware of the coordinator, not the other way around.

What is `coordinator.data`? I see the example in the docs uses it when creating entities, but does not explain why it is doing that or what it is used for... What methods am I expected to implement? What methods does HA call, and in what order?

Sorry I am just ranting right now. I feel like the documentation needs a lot of work, not just the coordinator section. And I'm not alone with this feeling. There was a whole reddit post about it https://www.reddit.com/r/homeassistant/comments/svg8s8/home_assistant_docs_is_a_nightmare_a_new_user/ .

Anyway, I just need to take a break and try to understand this after some rest.

bdraco replied:
You could move the zone discovery into `__init__.py` into a new function along with `open` so if it does fail then setup would be retried. Then store the results along side the data in `hass.data.setdefault(DOMAIN, {})[entry.entry_id]`  and then create a coordinator per zone (example https://github.com/home-assistant/core/blob/dev/homeassistant/components/lookin/__init__.py#L114)

If you get stuck, I can work up an example for you once the other comments are addressed

ssaenger replied:
I managed to figure it out. Thanks for the offer.

bdraco reviewed (changes requested):
Please see comments above

ssaenger committed:
Improve entity generation

ssaenger committed:
Implement coordinator

ssaenger commented:
Added coordinator to retrieve data and update my entities. I call `self.schedule_update_ha_state(True)` when a service is called (such as "set volume" to tell HA to pull data from the WS66i amp in order to update my states. However, sometimes there is a big delay between when `self.schedule_update_ha_state(True)` is called and when data is pulled, so there are times where there is a big delay from when the user changes a state to when lovelace updates the state. It seems like HA is throttling my calls to `self.schedule_update_ha_state(True)`. Unfortunately this makes for a bad user experience.

I still need to test shutting off my amp while integration is active...

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,127 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Soundavo WS66i 6-Zone Amplifier from a config entry."""
+    # Check if options was set. If not, initialize a dict for it
+    options: dict[str, dict[str, str]]
+    if not entry.options:
+        options = {CONF_SOURCES: INIT_OPTIONS_DEFAULT}
+        hass.config_entries.async_update_entry(entry, options=options)
+    else:
```
You can avoid this complexity by setting `options=` when you do `async_create_entry` in the config flow

bdraco commented on homeassistant/components/ws66i/__init__.py:
```diff
@@ -0,0 +1,127 @@
+    )
+
+    # Fetch initial data
+    await coordinator.async_refresh()
```
```suggestion
    await coordinator.async_config_entry_first_refresh()
```

bdraco commented on homeassistant/components/ws66i/coordinator.py:
```diff
@@ -0,0 +1,54 @@
+
+    async def _async_update_data(self) -> list[ZoneStatus]:
+        """Fetch data for each of the zones."""
+        data = []
+        data = await self.hass.async_add_executor_job(self._update_all_zones)
+
+        # HA will call my entity's _handle_coordinator_update()
+        # The data I pass back here can be accessed through coordinator.data.
+        return data
```
```suggestion
        return await self.hass.async_add_executor_job(self._update_all_zones)
```

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,208 @@
+) -> None:
+    """Set up the WS66i 6-zone amplifier platform from a config entry."""
+    ws66i_data: Ws66iData = hass.data[DOMAIN][config_entry.entry_id]
+    entities = []
+
+    for idx, zone_id in enumerate(ws66i_data.zones):
+        entities.append(
+            Ws66iZone(
+                device=ws66i_data.device,
+                ws66i_data=ws66i_data,
+                entry_id=config_entry.entry_id,
+                zone_id=zone_id,
+                idx=idx,
+                coordinator=ws66i_data.coordinator,
+            )
+        )
+
+    # Add them to HA
+    async_add_entities(entities)
```
```suggestion
    async_add_entities(
        Ws66iZone(
            device=ws66i_data.device,
            ws66i_data=ws66i_data,
            entry_id=config_entry.entry_id,
            zone_id=zone_id,
            idx=idx,
            coordinator=ws66i_data.coordinator,
        )
        for idx, zone_id in enumerate(ws66i_data.zones)
    )
```

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,208 @@
+        self._ws66i.set_mute(self._zone_id, mute)
+        self.schedule_update_ha_state(True)
+
+    def set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        self._ws66i.set_volume(self._zone_id, int(volume * 38))
+        self.schedule_update_ha_state(True)
```
```suggestion
    async def async_set_volume_level(self, volume):
        """Set volume level, range 0..1."""
        await self.hass.async_add_executor_job(self._ws66i.set_volume, self._zone_id, int(volume * 38))
        self.coordinator.async_request_refresh()
```

You can use `async_request_refresh` instead of `schedule_update_ha_state`

If you are reasonably sure the state change was successful, you could call also call `self._update_zone(zone_status)` with the new state and then call `self.async_write_ha_state()` to update the entity before doing that

ssaenger replied:
Calling self.async_write_ha_state() throws an exception in my service calls, such as def turn_off()
```
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Setting power to OFF for 0
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Getting state for zone 0. Power: off
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Getting state for zone 0. Power: off
2022-03-19 21:21:05 INFO (SyncWorker_0) [homeassistant.components.ws66i.media_player] Getting state for zone 0. Power: off
2022-03-19 21:21:05 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [139905263015728] Non-thread-safe operation invoked on an event loop other than the current one
Traceback (most recent call last):
  File "/workspaces/core/homeassistant/components/websocket_api/commands.py", line 192, in handle_call_service
    await hass.services.async_call(
  File "/workspaces/core/homeassistant/core.py", line 1636, in async_call
    task.result()
  File "/workspaces/core/homeassistant/core.py", line 1673, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/workspaces/core/homeassistant/helpers/entity_component.py", line 204, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/workspaces/core/homeassistant/helpers/service.py", line 668, in entity_service_call
    future.result()  # pop exception if have
  File "/workspaces/core/homeassistant/helpers/entity.py", line 931, in async_request_call
    await coro
  File "/workspaces/core/homeassistant/helpers/service.py", line 705, in _handle_entity_call
    await result
  File "/workspaces/core/homeassistant/components/media_player/__init__.py", line 649, in async_turn_off
    await self.hass.async_add_executor_job(self.turn_off)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/workspaces/core/homeassistant/components/ws66i/media_player.py", line 188, in turn_off
    self.async_write_ha_state()
  File "/workspaces/core/homeassistant/helpers/entity.py", line 539, in async_write_ha_state
    self._async_write_ha_state()
  File "/workspaces/core/homeassistant/helpers/entity.py", line 663, in _async_write_ha_state
    self.hass.states.async_set(
  File "/workspaces/core/homeassistant/core.py", line 1373, in async_set
    self._bus.async_fire(
  File "/workspaces/core/homeassistant/core.py", line 862, in async_fire
    self._hass.async_add_hass_job(job, event)
  File "/workspaces/core/homeassistant/core.py", line 446, in async_add_hass_job
    self.loop.call_soon(hassjob.target, *args)
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 748, in call_soon
    self._check_thread()
  File "/usr/local/lib/python3.9/asyncio/base_events.py", line 785, in _check_thread
    raise RuntimeError(
RuntimeError: Non-thread-safe operation invoked on an event loop other than the current one
```
In the stack trace, you will see "self.async_write_ha_state()" is called in my ws66i entity class. Is this a bug?

bdraco replied:
You need to rename them `async_` .. and make them co-routines and wrap the call in `async_add_executor_job` to be able to use `async_write_ha_state`.  

How this works is that `async def async_set_volume_level(self, volume):` is calling `set_volume_level` under the hood
https://github.com/home-assistant/core/blob/dev/homeassistant/components/media_player/__init__.py#L665

By implementing `async_set_volume_level` we need to do the executor call ourselves.

ssaenger replied:
So coordinator updates run on the MainThread, but service calls run on a syncWorker. Calling async_write_ha_state() is not allowed from this thread. I'm going to look into how to circumvent this.

ssaenger replied:
Okay, the solution is to make all my service calls asynchronous, as mentioned in the documentation https://developers.home-assistant.io/docs/integration_fetching_data/#push-vs-poll

ssaenger replied:
Thanks for the response. I didn't refresh the page when I wrote my two other comments after your reply, so I didn't see it.

ssaenger committed:
Made options fields required, retry connection on failed attempts, use ZoneStatus for attributes

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,224 @@
+        # Parent will notify HA that our fields were updated
+        super()._handle_coordinator_update()
+
+    @property
+    def state(self) -> str:
+        """State of the player."""
+        return STATE_ON if self._status.power else STATE_OFF
+
+    @property
+    def volume_level(self) -> float:
+        """Volume level of the media player (0..1)."""
+        return self._status.volume / float(MAX_VAL)
+
+    @property
+    def is_volume_muted(self) -> bool:
+        """Boolean if volume is currently muted."""
+        return self._status.mute
+
+    @property
+    def media_title(self) -> str:
+        """Title of current playing media."""
+        return self._ws66i_data.sources.id_name[self._status.source]
+
+    @property
+    def source(self) -> str:
+        """Name of the current input source."""
+        # Source name is found from the dataclass
+        return self._ws66i_data.sources.id_name[self._status.source]
```
You could drop these and set them as `self._attr_X` in `_handle_coordinator_update` instead

bdraco replied:
Example https://github.com/bdraco/home-assistant/commit/e777ed57ef23b7aef9dffb21732df0dc522201d2

ssaenger replied:
I like it.

ssaenger commented:
I am confident that commands sent off to the WS66i amp will be accepted, so my entity class will update its state and alert HA by calling `self.async_write_ha_state()`. This will provide immediate feedback to the end-user, and won't poll the amp each time the user makes a change through HA.

Since pyws66i returns a ZoneStatus class, my entities will reference that instead of copying those values locally. This means I brought back some `@property` attributes as they will need to read from the ZoneStatus class.

I've added coordinator handling to close the WS66i connection if the data pipeline breaks (such as the amp powered off) and on subsequent coordinator updates, it will retry to establish a connection.

I made the fields in the options flow required. If the user were to not provide a value in a config flow, that user would not be able to select a source through HA. This is undesired and required additional error checking in a few other places. This is especially problematic if the user selected a source via the keypad that wasn't configured in the config flow. By making them mandatory, the user has the freedom to select all available sources and some error checking was no longer needed, simplifying code.

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,224 @@
+
+    async def async_restore(self):
+        """Restore saved state."""
+        if self._snapshot:
```
```suggestion
        if not self._snapshot:
            raise HomeAssistantError("There is no snapshot to restore")
```

bdraco replied:
Then outdent below

ssaenger committed:
Refactor WS66i entity properties, raise HomeAssistantError on restore service if no snapshot

ssaenger commented on homeassistant/components/ws66i/coordinator.py:
```diff
@@ -1,4 +1,6 @@
 """Coordinator for WS66i."""
+from __future__ import annotations
```
Oops I left this in. It was intended for some work that I scrapped.

ssaenger commented:
I tried making the restore service more robust by remembering what the snapshot status was between HA restarts, however the data in `hass.data` is wiped out between restarts. Is there any place to store data persistently for integrations?

bdraco commented:
> I tried making the restore service more robust by remembering what the snapshot status was between HA restarts, however the data in `hass.data` is wiped out between restarts. Is there any place to store data persistently for integrations?

 I suggest removing the services for now and working through them in a followup PR. 

https://developers.home-assistant.io/docs/creating_component_code_review?_highlight=small&_highlight=pos#5-make-your-pull-request-as-small-as-possible

ssaenger commented:
Sorry, I was getting ahead of myself. I was trying to continue to make improvements during a PR and that just makes it more difficult for you, the reviewer.

About custom services, these services should stay however; they are vital components to the integration. A basic automation will, for example, listen for an event, such as a doorbell rang event, snapshot the zones, play a notification indicating someone is at the door, and then restore the zones. In this use case, the snapshot will no longer be needed because a new snapshot will be taken whenever a notification needs to be played. Come to think of it, snapshots don't need to be in non-volatile memory at all.

I won't be making any more changes. In a future PR, I'll look into discovery. I hope this is acceptable.

bdraco commented on homeassistant/components/ws66i/coordinator.py:
```diff
@@ -0,0 +1,66 @@
+
+    def _update_all_zones(self) -> list[ZoneStatus]:
+        """Fetch data for each of the zones."""
+        if self._con_broken:
+            # Try to re-establish a connection
```
Can this be moved into the upstream library? Usually we don't want Home Assistant to have to know about he protocol details if it can be avoided (such as the need to re-establish the connection).  Normally we would expect calling something like 'zone_status' would reconnect if needed

ssaenger replied:
I do need to update the upstream library to fix a few minor things, and I think you are right about `zone_status`, however I am not sure if it should reconnect. If `zone_status` could reconnect, then the `open` method almost seems pointless. If applications skip calling `open`, then they might not think about calling `close` when they are done. The way I see it, `open` needs to be called before anything can be done and `close` needs to be called when it is done. I do think though that `zone_status` should raise a `ConnectionError` when the connection is lost. This will alert the application to call `open` again to re-establish it. I don't think the application should know to call `close` like I am doing a few lines down, so the upstream library should take care of that.

If you think `zone_status` should reconnect regardless, I have no issue making that change.

bdraco replied:
It seems like it should auto reconnect if it's in an open state, but not if it's in a closed state.

Raising an exception and handling it is ok as well if there is a use case where it would be a bad idea to auto reconnect

ssaenger replied:
Done.

ssaenger committed:
Update to pyws66i v1.1

ssaenger committed:
Add quality scale of silver to manifest

ssaenger commented:
Updated to use pyws66i v1.1.

Reviewed the [quality scale](https://developers.home-assistant.io/docs/integration_quality_scale_index/) and upgraded my integration to Silver.

bdraco commented on tests/components/ws66i/test_config_flow.py:
```diff
@@ -0,0 +1,147 @@
+    )
+    config_entry.add_to_hass(hass)
+
+    with patch("homeassistant.components.ws66i.async_setup_entry", return_value=True):
```
For the options flow we need to setup the integration so we should patch the lib instead

ssaenger replied:
Is this what you have in mind?

```
-    with patch("homeassistant.components.ws66i.async_setup_entry", return_value=True):
+    with patch("homeassistant.components.ws66i.get_ws66i") as mock_ws66i:
+        ws66i_instance = mock_ws66i.return_value
+        ws66i_instance.zone_status.return_value = AttrDict(
+                power=True, volume=0, mute=True, source=1, treble=0, bass=0, balance=10
+            )
         assert await hass.config_entries.async_setup(config_entry.entry_id)
         await hass.async_block_till_done()
```
100% code coverage on all files.

bdraco commented:
LGTM once https://github.com/home-assistant/core/pull/56094#discussion_r842340040 is addressed

ssaenger committed:
Update config_flow test

MartinHjelmare commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,213 @@
+        status = self._status
+        sources = self._ws66i_data.sources.id_name
+        self._attr_state = STATE_ON if status.power else STATE_OFF
+        self._attr_volume_level = status.volume / float(MAX_VOL)
```
Maybe round the result?

ssaenger replied:
I don't think so. Volume level should be between 0 and 1.

MartinHjelmare commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,213 @@
+    async def async_set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        await self.hass.async_add_executor_job(
+            self._ws66i.set_volume, self._zone_id, int(volume * MAX_VOL)
```
Maybe use `round` instead of `int`.

ssaenger replied:
The 2nd parameter of `set_volume()` expects an int. `round()` returns a float. I think keeping it an int is good.

MartinHjelmare commented on homeassistant/components/ws66i/models.py:
```diff
@@ -0,0 +1,30 @@
+
+from .coordinator import Ws66iDataUpdateCoordinator
+
+# A dataclass is basically a struct in C/C++
```
We don't need to document Python features. They are documented in the Python docs.

ssaenger replied:
I was confused what dataclasses were. Once I made the connection to a data struct found in C, it made sense and I put it in there for my own benefit, but I'll remove the comment.

MartinHjelmare commented on tests/components/ws66i/test_media_player.py:
```diff
@@ -0,0 +1,692 @@
+    await hass.services.async_call(DOMAIN, name, service_data=data, blocking=True)
+
+
+async def test_cannot_connect(hass):
```
This test should move to a `test_init.py` module.

MartinHjelmare commented on tests/components/ws66i/test_media_player.py:
```diff
@@ -0,0 +1,692 @@
+        hass, SERVICE_SELECT_SOURCE, {"entity_id": ZONE_1_ID, "source": "three"}
+    )
+
+    # await coordinator.async_refresh()
```
Commented code. Please remove it.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!

bdraco commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,213 @@
+PARALLEL_UPDATES = 1
+
+SUPPORT_WS66I = (
+    SUPPORT_VOLUME_MUTE
```
I fixed this in https://github.com/home-assistant/core/pull/71553 since it was causing pylint to warn

ssaenger commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,213 @@
+    # Set up services
+    platform = async_get_current_platform()
+
+    platform.async_register_entity_service(
+        SERVICE_SNAPSHOT,
+        {},
+        "snapshot",
+    )
+
+    platform.async_register_entity_service(
+        SERVICE_RESTORE,
+        {},
+        "async_restore",
+    )
```
I have these two services because I modeled this integration after the monoprice integration.

I was just playing with scenes after you mentioned this. I couldn't see a way for the scene to save the state of the entity at the time of an automation being triggered. A use case of the snapshots/restore is:

1. User is listening to background music in the kitchen on source 3, volume 10 (arbitrary values)
2. Doorbell is rung
3. An automation is triggered
   a. The snapshot service is called to save the state
   b. A scene is loaded, which sets the source to 4, volume to 20
   c. A media file is played ("ding-dong" notification sound)
   d. The restore service is called to revert state back to source 3, volume 10

The user is notified that someone is at the door, and the background music resumes. Can this behavior be replicated without these services?

frenck replied:
That is exactly the services scene provides and was created for. Scenes allow you to capture states and restore states (it actually uses state restore under the hood).

For example, you could create a snapshot of a speaker, before doing anything:

```yaml
- service: scene.create
  data:
    scene_id: snapshot_of_sonos_office
    snapshot_entities:
      - media_player.sonos_office
```

Next, do the stuff one wants... and after that sequence is done, restore the previous snapped scene:

```yaml
- service: scene.turn_on
  target:
    entity_id: scene.snapshot_of_sonos_office
```

ssaenger replied:
Perfect! Thanks for the thorough response. Those custom services are not needed. I'll remove them and add that change to the PR that @MartinHjelmare is requesting.

ssaenger commented on homeassistant/components/ws66i/coordinator.py:
```diff
@@ -0,0 +1,53 @@
+
+            data.append(data_zone)
+
+        # HA will call my entity's _handle_coordinator_update()
```
Coordinators was a concept I struggled immensely. I did not understand the flow of coordinators, such as, "after this method is called, what will it call next?" I added this comment to remind me what HA will do next. I'll remove the comment though.

ssaenger commented on homeassistant/components/ws66i/media_player.py:
```diff
@@ -0,0 +1,213 @@
+        self._attr_source_list = ws66i_data.sources.name_list
+        self._attr_unique_id = f"{entry_id}_{self._zone_id}"
+        self._attr_name = f"Zone {self._zone_id}"
+        self._attr_supported_features = SUPPORT_WS66I
```
This was resolved in [#71553](https://github.com/home-assistant/core/pull/71553).

ssaenger commented on homeassistant/components/ws66i/strings.json:
```diff
@@ -0,0 +1,34 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
Do I update the translation file as well?

MartinHjelmare replied:
You can run the translations develop script if you want to test the flow manually. It's not needed to update the translation files for merging a PR. Our infrastructure will update the translation files separately by fetching from Lokalise.

ssaenger commented:
@MartinHjelmare , yours and Franck's requested changes are in PR https://github.com/home-assistant/core/pull/71717
